lib/THC/THCTensorMathMagma.cu

```diff
@@ -0,0 +1,472 @@
+  THCudaTensor_copyArray2d(state, ru_, ru_data, m, j);
+  THCudaTensor_copyArray1d(state, rs_, rs_data, k);
+  THCudaTensor_copyArray2d(state, ra_, a_data,  m, n);
+
+  magma_free_pinned(work_data);
+  magma_free_pinned(rv_data);
+  magma_free_pinned(ru_data);
+  magma_free_pinned(rs_data);
+  magma_free_pinned(a_data);
+#else
+  THError(NoMagma(gesvd2));
+#endif
+}
+
+void THCudaTensor_getri(THCState *state, THCudaTensor *ra_, THCudaTensor *a)
+{
+#ifdef USE_MAGMA
+  THArgCheck(a->nDimension == 2, 2, "A should be 2 dimensional");
+  THArgCheck(a->size[0] == a->size[1], 2, "A should be square");
+
+  int info;
+  int n = a->size[0];
+  int lwork = n * magma_get_sgetri_nb(n);
+
+  THCudaTensor *input = THCudaTensor_newColumnMajor(state, ra_, a);
+  float *input_data = THCudaTensor_data(state, input);
+
+  int *ipiv = th_magma_imalloc_pinned(n);
+
+  THCudaTensor *work = THCudaTensor_newWithSize1d(state, lwork);
+  float *work_data = THCudaTensor_data(state, work);
+
+  // Run LU
+  magma_sgetrf_gpu(n, n, input_data, n, ipiv, &info);
+  if (info > 0)
+    THError("MAGMA getrf : U(%d,%d) is 0, U is singular", info, info);
+  else if (info < 0)
+    THError("MAGMA getrf : Argument %d : illegal value", -info);
+
+  // Inverse
+  magma_sgetri_gpu(n, input_data, n, ipiv, work_data, lwork, &info);
+  if (info > 0)
+    THError("MAGMA getri : U(%d,%d) is 0, U is singular", info, info);
+  else if (info < 0)
+    THError("MAGMA getri : Argument %d : illegal value", -info);
+
+  THCudaTensor_free(state, work);
+  magma_free_pinned(ipiv);
+  THCudaTensor_freeCopyTo(state, input, ra_);
+#else
+  THError(NoMagma(getri));
+#endif
+}
+
+__global__ void THCudaTensor_copyUpperSymmetric(float *input, int n, int len)
+{
+  const int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx < len) {
+    const int r = idx % n;
+    const int c = idx / n;
+    if (r > c) {
+      input[idx] = input[r*n + c];
+    }
+  }
+}
+
+void THCudaTensor_potri(THCState *state, THCudaTensor *ra_, THCudaTensor *a)
+{
+#ifdef USE_MAGMA
+  THArgCheck(a->nDimension == 2, 2, "A should be 2 dimensional");
+  THArgCheck(a->size[0] == a->size[1], 2, "A should be square");
+
+  int n = a->size[0];
+
+  THCudaTensor *input = THCudaTensor_newColumnMajor(state, ra_, a);
+  float *input_data = THCudaTensor_data(state, input);
+
+  int info;
+  magma_spotrf_gpu(MagmaUpper, n, input_data, n, &info);
+  if (info > 0)
+    THError("MAGMA potrf : A(%d,%d) is 0, A cannot be factorized", info, info);
+  else if (info < 0)
+    THError("MAGMA potrf : Argument %d : illegal value", -info);
+
+  magma_spotri_gpu(MagmaUpper, n, input_data, n, &info);
+  if (info > 0)
+    THError("MAGMA potri : A(%d,%d) is 0, A cannot be factorized", info, info);
+  else if (info < 0)
+    THError("MAGMA potri : Argument %d : illegal value", -info);
+
+  const int len = n*n;
+  cudaStream_t stream = THCState_getCurrentStream(state);
+  THCudaTensor_copyUpperSymmetric<<<(len+127/128), 128, 0, stream>>>(input_data, n, len);
```

Comment:
Make sure you don't launch more than the maximum number of thread blocks. You can query it from `state->deviceProperties[currentDevice].maxGridSize[0]`.